File: src/main/java/Solution_148_SortList.java
public class Solution_148_SortList {
    public static void main(String[] args) {
        var node1 = new ListNode2(-1);
        var node2 = new ListNode2(5);
        var node3 = new ListNode2(3);
        var node4 = new ListNode2(4);
        var node5 = new ListNode2(0);
        node1.setNext(node2);
        node2.setNext(node3);
        node3.setNext(node4);
        node4.setNext(node5);
        sortList(node1);
    }

    private static ListNode2 sortList(ListNode2 head) {
        if (head == null || head.next == null) return head;

        // 1. Найти середину
        var mid = getMid(head);
        var rightMid = mid.next;
        mid.next = null;

        // 2. Рекурсивная сортировка
        var left = sortList(head);
        var right = sortList(rightMid);

        // 3. Слияние
        return merge(left, right);
    }

    private static ListNode2 getMid(ListNode2 node) {
        var slow = node;
        var fast = node.next;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    private static ListNode2 merge(ListNode2 list1, ListNode2 list2) {
        var dummy = new ListNode2(0);
        var tail = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                tail.next = list1;
                list1 = list1.next;
            } else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }

        tail.next = list1 != null ? list1 : list2;

        return dummy.next;
    }
}

File: src/main/kotlin/Solution_148_SortList.kt
/**
 * Given the head of a linked list, return the list after sorting it in ascending order.

 *
 * Example 1:
 *
 *
 * Input: head = [4,2,1,3]
 * Output: [1,2,3,4]
 * Example 2:
 *
 *
 * Input: head = [-1,5,3,4,0]
 * Output: [-1,0,3,4,5]
 * Example 3:
 *
 * Input: head = []
 * Output: []
 *
 *
 * Constraints:
 *
 * The number of nodes in the list is in the range [0, 5 * 104].
 * -105 <= Node.val <= 105
 *
 */
fun main() {
    val node1 = ListNode(-1)
    val node2 = ListNode(5)
    val node3 = ListNode(3)
    val node4 = ListNode(4)
    val node5 = ListNode(0)
    val node6 = ListNode(6)
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    node5.next = node6
    sortList(node1)
}

/**
 * Merge Sort (сортировка слиянием)
 * Идея решения
 * Разделить список на 2 половины (через slow/fast указатели)
 * Рекурсивно отсортировать обе части
 * Слить два отсортированных списка
 *
 * Сложность
 * Время: O(n log n)
 * Память: O(log n) (из-за рекурсии)
 */
fun sortList(head: ListNode?): ListNode? {
    if (head?.next == null) return head

    // 1. Делим список пополам
    val mid = getMid(head)
    val rightHead = mid.next
    // Чтобы разорвать список на два отдельных списка, иначе будет зацикливание
    mid.next = null

    // 2. Рекурсивно сортируем
    val left = sortList(head)
    val right = sortList(rightHead)

    // 3. Сливаем
    return merge(left, right)
}

/**
 * 1. Как найти середину?
 * var slow = head
 * var fast = head.next
 * slow идёт на 1 шаг
 * fast на 2 шага
 * ➡️ когда fast дошёл до конца — slow в середине
 */
private fun getMid(head: ListNode): ListNode {
    var slow = head
    var fast = head.next

    while (fast != null && fast.next != null) {
        slow = slow.next!!
        fast = fast.next?.next
    }

    return slow
}

private fun merge(l1: ListNode?, l2: ListNode?): ListNode? {
    val dummy = ListNode(0)
    var tail = dummy

    var left = l1
    var right = l2

    while (left != null && right != null) {
        if (left.`val` < right.`val`) {
            tail.next = left
            left = left.next
        } else {
            tail.next = right
            right = right.next
        }
        tail = tail.next!!
    }

    tail.next = left ?: right

    return dummy.next
}